Split "Locomotor" from Mobile

reaperrr opened the pull request:
This moves a lot of stuff from `Mobile` to `Locomotor` world trait(s).

The idea is to keep the number of locomotors limited and - in the future - cache/pre-calculate as many things as possible on them, to save performance and make further movement code refactors easier.

reaperrr committed:
Make RA use Locomotor

PeterAntal commented:
I'd be open to try help if I can better understand the intent. 

Can you explain the desired relationship of Mobile, MobileInfo and MovementClassInfo in your design change here?

PeterAntal commented:
Are you manually testing, or have you been running the "make test" flow? I'm seeing a bunch non-blocking load failures getting logged, when instantiating the new MovementClassInfo at load time. These might be happening upstream of your crashes.

_Testing mods...
Testing Tiberian Sun mod MiniYAML...
Testing mod: Tiberian Sun
OpenRA.Utility(1,1): Error: Actor world is not constructible; failure: ActorInfo("world") failed to initialize because of the following:
Missing:
OpenRA.Mods.Common.Traits.MovementClassInfo
Unresolved:
OpenRA.Mods.Common.Traits.PathFinderInfo: { OpenRA.Mods.Common.Traits.MovementClassInfo }

OpenRA.Utility(1,1): Error: Actor type `weedguy` consumes conditions that are not granted: inside-tunnel
..._

reaperrr commented:
@PeterAntal I only migrated RA so far, so test errors in the other mods are to be expected.
RA passes make test, but still crashes at launch with the exception linked in the OP.

The idea here is to use a limited number of MovementClass (a different name like MovementType might be better, but this is what I went with, for now) traits residing on the world actor that cache their terrain speed and cost  etc. at game start, which Mobile then only needs to access, instead of doing this for every actor with Mobile separately.

This is meant to be the first step of #14830. See also the discussions on IRC (log) from yesterday and around 19-20 days ago, or try to get hold of @pchote or @chrisforbes here or on IRC for any additional technical questions (I understand the concept, but I'm not good at explaining the details).

reaperrr committed:
Make TD use Locomotor

reaperrr committed:
Make D2k use Locomotor

reaperrr committed:
Make TS use Locomotor

reaperrr commented:
Updated.

There are still issues with the TS-specific stuff (at least with Tunnel-, Jumpjet- and SubterraneanConditions), but the other 3 mods now pass all tests and seem to work fine.

PeterAntal commented:
Got it - so, to paraphrase, it sounds like the general aspiration is:
1/2- pre-compute a graphs summarizing the long distance pathfinding model, representing each kind of movement
3-Exercise long distance navigation through the model in place of present scheme.

I'll see if I can get context on what should be implementing traits of your new MovementClass.

PeterAntal commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -145,234 +50,20 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
 		public bool CanEnterCell(World world, Actor self, CPos cell, Actor ignoreActor = null, bool checkTransientActors = true)
 		{
-			if (MovementCostForCell(world, cell) == int.MaxValue)
+			var movementClassInfo = world.WorldActor.TraitsImplementing<MovementClass>().Single(mc => mc.Info.Class == MovementClass).Info;
```
The worldActor does not itself include MovementClass Trait, which the mobile instances are looking for.

Are you sure a trait on the world the right mechanism for obtaining an instance?

Seems like you want something of a instance management mechanism in relation to the world actor, to request instantiation of the MovementClass, if it doesn't already exist(the world probably shouldn't need to declare what specific Movement Class instances are needed).

pchote replied:
OpenRA uses traits on the world to hold various types of global state, including palettes, faction definitions, and references for the in-game UI.  Defining the movement types on the world also matches up with the original games (which called them "locomotors"), and makes it more explicit that these are things that should be limited in number.

pchote replied:
But please do this as `world.WorldActor.Info.TraitInfos<MovementClassInfo>().Single(mc => mc.Class == MovementClass)` instead of going via the actual world actor.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -22,49 +22,11 @@
-		public readonly int WaitAverage = 5;
-
-		public readonly int WaitSpread = 2;
+		public readonly string MovementClass = "default";
```
It would be good to define a `[MovementClassReference]` attribute that can be linted in the same way as actor, weapon, sequence, etc references.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -145,234 +50,20 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
-		}
+		public int GetInitialFacing() { return InitialFacing; }
 
+		public bool SharesCell = false;
```
It would be cleaner to use the `IRulesetLoaded` to store a reference to the `MovementClassInfo`, the same way that we do for weapons in `ArmamentInfo`.  This can then be removed.

GraionDilach replied:
Shifting this towards movementclass might imply the plumbing/possibility for multi-cell units which sounds out-of-scope on this run to me though.

pchote replied:
I'm not sure what you mean.  This is about code quality, not feature changes.

GraionDilach replied:
I understand that aspect, but not all modders might follow us in that regard though.

pchote replied:
This variable holds a cached value of the field of the same name which has already moved to `MovementClassInfo`.  It is overwritten with that value whenever `CanEnterCell` is called.  If a modder tries to set this from yaml the game will ignore it.

GraionDilach commented:
I'm not really pleased with the idea to tie crushing abilities to the locomotor tbh.

pchote commented:
The ability to crush is one of the most important parameters in the pathing state.  It needs to move.

GraionDilach commented:
Well, Crushable is conditional already so I guess this can go that way then and maybe it's just me who's seeing issues atm.

reaperrr commented:
Updated,  managed to work out all the initial kinks (even tunnels, jumpjets and subterranean units work now). I think this looks good enough now that I can at least remove the help request label.

One thing I'd like to get feedback on, though:
Should we stick to `MovementClass`, or go for `MovementType` or just `Locomotor` (like in TS/RA2)?
I'm asking because I noticed there are some places in the Pathfinder code that already used the term `MovementClass` in a slightly different context.

pchote commented:
My preference is for `Locomotor` because `Movement*` is misleading (this logic only applies to `Mobile`, not the other movement types).

MustaphaTR commented:
I don't think what this does and what Locomotor in original games does is the same thing. This is more like the [SpeedType](https://www.modenc.renegadeprojects.com/SpeedType). Not sure if is a good name, but i don't think Locomotor is good either.

For reference, here is Locomotor's ModEnc page: https://www.modenc.renegadeprojects.com/Locomotors

reaperrr commented:
> I don't think what this does and what Locomotor in original games does is the same thing.
This is more like the SpeedType. Not sure if is a good name, but i don't think Locomotor is good either.

That may be the case right now, but there are at least two things to consider:
- making the trait name more accurate would make it longer, too (`GroundMovementPropertyClass`, `GroundMovementBehaviorType` or something along these lines would be a bit long)
- the idea isn't to leave the trait like it is now (which already goes quite beyond the Westwood `SpeedType`), but rather to move more stuff from `Mobile` here later, turning `Mobile` more into a simpler, "carries some properties, some interfaces, and calls some external methods" kind of trait, instead of doing most locomotor-like stuff by itself.

reaperrr committed:
Split Locomotor trait from Mobile

Add GrantConditionOn*Layer traits

This allows to
- drop some booleans from Locomotor
- drop a good part of the subterranean- and jumpjet-specific code/hacks from Mobile
- grant more than 1 condition per layer type (via multiple traits)
- easily add more traits of this kind for other layers

reaperrr committed:
Add Locomotor lint check

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -50,6 +50,17 @@ public interface INotifySold
+	[RequireExplicitImplementation]
+	public interface INotifyChangedCustomLayer
+	{
+		void JumpjetLaunch(Actor self);
```
Can we please simplify this down to a single `CustomLayerChanged(CustomMovementLayerType old, CustomMovementLayerType new)`?

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCustomLayer.cs:
```diff
@@ -0,0 +1,89 @@
+		public object Create(ActorInitializer init) { return new GrantConditionOnCustomLayer(init, this); }
+	}
+
+	public class GrantConditionOnCustomLayer : INotifyCreated, INotifyChangedCustomLayer
```
and then split this into separate traits for Tunnel/Subterranean/Jumpjet.  You could define a single abstract class that is given the `CustomLayerMovementType` to check in its ctor from the subclass.  The `CustomLayerChanged` method in each can then check this value and grant/revoke the condition when needed.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -429,28 +128,36 @@ public void SetLocation(CPos from, SubCell fromSub, CPos to, SubCell toSub)
-				tunnelToken = conditionManager.RevokeCondition(self, tunnelToken);
+			if (toCell.Layer == CustomMovementLayerType.Tunnel && fromCell.Layer != CustomMovementLayerType.Tunnel)
+				foreach (var n in notifyCustomLayerChange)
+					n.TunnelEnter(self);
```
A lot of the logic here can then go away.

pchote commented on OpenRA.Mods.Common/Traits/World/MovementClass.cs:
```diff
@@ -0,0 +1,341 @@
+		public readonly bool MoveIntoShroud = true;
+
+		[Desc("Can this unit move underground?")]
+		public readonly bool Subterranean = false;
```
Can you please define subclasses for `SubterraneanMovementClassInfo` and the other custom types?

The checks for `if (mc.Subterranean)` can then be replaced by `if (mc is SubterraneanMovementClassInfo)` or
```
var smc = mc as SubterraneanMovementClassInfo;
if (smc != null)
```
if the Subterranean specific info is needed.

pchote reviewed (commented):
Would you mind splitting this up a bit further?  These changes should go a long way towards allowing modders to define their own custom movement types without having to modify the common logic.

reaperrr commented:
Updated.

reaperrr commented:
Updated again.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnLayer.cs:
```diff
@@ -0,0 +1,71 @@
+		protected ConditionManager conditionManager;
+		protected int conditionToken = ConditionManager.InvalidConditionToken;
+
+		protected virtual byte ValidLayerType { get; set; }
```
Can you make this a `protected readonly byte validLayerType` and then assign it from the ctor here?

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnLayer.cs:
```diff
@@ -0,0 +1,71 @@
+
+		protected virtual byte ValidLayerType { get; set; }
+
+		public GrantConditionOnLayer(Actor self, GrantConditionOnLayerInfo info)
```
(passing the layer type from here)

reaperrr replied:
With "from here", do you mean literally this line? If yes, why?

pchote replied:
```csharp
public GrantConditionOnLayer(Actor self, GrantConditionOnLayerInfo info, byte layerType)
	: base(info)
{
	this.layerType = layerType;
}
```

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnJumpjetLayer.cs:
```diff
@@ -0,0 +1,39 @@
+			ValidLayerType = CustomMovementLayerType.Jumpjet;
+		}
+
+		protected override void UpdateConditions(Actor self, byte oldLayer, byte newLayer)
```
Can you elaborate on why this needs to be special cased?

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -145,240 +52,33 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
-		{
-			return MovementCostForCell(world, TilesetTerrainInfo[world.Map.Rules.TileSet], cell);
-		}
+		public LocomotorInfo LinkedLocomotorInfo;
```
Can we call this just `LocomotorInfo`?  This should also be `{ get; private set; }` to avoid shenanigans.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -145,240 +52,33 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
-
-			var index = cell.Layer == 0 ? world.Map.GetTerrainIndex(cell) :
-				world.GetCustomMovementLayers()[cell.Layer].GetTerrainIndex(cell);
+			var world = rules.Actors["world"];
```
This is only used once, so just inline it.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -429,28 +128,15 @@ public void SetLocation(CPos from, SubCell fromSub, CPos to, SubCell toSub)
-			if (toCell.Layer == CustomMovementLayerType.Jumpjet && conditionManager != null &&
-					!string.IsNullOrEmpty(Info.JumpjetCondition) && jumpjetToken == ConditionManager.InvalidConditionToken)
-				jumpjetToken = conditionManager.GrantCondition(self, Info.JumpjetCondition);
+			// Granting the jumpjet condition as soon as the actor starts leaving the ground layer already happened above
```
Is there a way we can move this into `GrantConditionOnJumpjetLayer`?

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -701,15 +389,21 @@ public void EnteringCell(Actor self)
-			if (fromCell.Layer != CustomMovementLayerType.Jumpjet && toCell.Layer != CustomMovementLayerType.Jumpjet && jumpjetToken != ConditionManager.InvalidConditionToken)
-				jumpjetToken = conditionManager.RevokeCondition(self, jumpjetToken);
+			var isJumpjet = Info.LinkedLocomotorInfo is JumpjetLocomotorInfo;
+			if (isJumpjet && jumpjetInAir && fromCell.Layer != CustomMovementLayerType.Jumpjet && toCell.Layer != CustomMovementLayerType.Jumpjet)
```
... and this too.  This may need a new `INotifyFinishedMoving` interface?

pchote commented on OpenRA.Mods.Common/Traits/World/PathFinder.cs:
```diff
@@ -86,6 +88,7 @@ public List<CPos> FindUnitPath(CPos source, CPos target, Actor self, Actor ignor
 		public List<CPos> FindUnitPathToRange(CPos source, SubCell srcSub, WPos target, WDist range, Actor self)
 		{
 			var mi = self.Info.TraitInfo<MobileInfo>();
+			var locoInfo = mi.LinkedLocomotorInfo;
```
Can we make this `li` for consistency with `mi` and other uses of `li`?

pchote commented on OpenRA.Mods.Common/Traits/World/SubterraneanActorLayer.cs:
```diff
@@ -71,28 +71,31 @@ WPos ICustomMovementLayer.CenterOfCell(CPos cell)
 		{
 			var terrainType = map.GetTerrainInfo(cell).Type;
-			if (!mi.SubterraneanTransitionTerrainTypes.Contains(terrainType) && mi.SubterraneanTransitionTerrainTypes.Any())
+			var loco = li as SubterraneanLocomotorInfo;
```
Can we use `sli` for consistency?

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -310,9 +316,9 @@ public interface ICustomMovementLayer
-		bool EnabledForActor(ActorInfo a, MobileInfo mi);
-		int EntryMovementCost(ActorInfo a, MobileInfo mi, CPos cell);
-		int ExitMovementCost(ActorInfo a, MobileInfo mi, CPos cell);
+		bool EnabledForActor(ActorInfo a, LocomotorInfo loco);
```
`loco` vs `li` consistency

pchote commented on mods/cnc/rules/world.yaml:
```diff
@@ -13,6 +13,126 @@
 		FogVariants: typea, typeb, typec, typed
 		OverrideFullShroud: full
 		OverrideFullFog: full
+	Locomotor@default:
```
`@DEFAULT` for consistency please.

reaperrr replied:
Maybe we should actually remove these, to prevent modders (and us) from accidentally forgetting to give all mobile actors proper locomotors (and then not even noticing because the actors silently use those fallbacks).

reaperrr replied:
Removing them actually turned out to be a good idea, caught a typo that way. Also added YamlExceptions at rules loading time, because mismatches crash the utility before the lint rule can tell you the problem (which makes me wonder whether those exceptions make the lint rule obsolete).

pchote commented on mods/d2k/rules/world.yaml:
```diff
@@ -15,6 +15,74 @@
 		OverrideFullShroud: shroudfull
 		OverrideFullFog: fogfull
 		ShroudBlend: Multiply
+	Locomotor@default:
```
Likewise here.

pchote commented on mods/ra/rules/world.yaml:
```diff
@@ -8,6 +8,96 @@
 		DefeatMusic: map
 	TerrainGeometryOverlay:
 	DebugVisualizations:
+	Locomotor@default:
```
and here.

pchote reviewed (commented):
Another quick pass of comments:

reaperrr commented:
Updated.

Disclaimer: I know this would be easier to review if https://github.com/OpenRA/OpenRA/pull/14893/commits/3c4f5882611c06b73e0c18f8c36bae096098c49f was split into multiple commits, but that would make it pretty hard to apply all these fixup/change requests without messing up during rebase.

reaperrr commented:
I think it would be good to get this into the next playtest.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -22,49 +22,13 @@
-		public readonly int WaitSpread = 2;
+		[Desc("Which Locomotor does this trait use. Must be defined on the World actor.")]
+		[LocomotorReference]
+		public readonly string Locomotor = "default";
```
I wonder if we don't just want to `Fieldloader.Require` this and leave the default empty.

abcdefg30 commented on mods/cnc/rules/world.yaml:
```diff
@@ -13,6 +13,117 @@
+			Road: 100
+			Bridge: 100
+			Tiberium: 90
+				PathingCost: 90
```
Now that we don't inherit the value of 300 from infantry (foot locomotor) anymore, does it still make sense to define these explicit?

abcdefg30 commented on mods/ts/rules/world.yaml:
```diff
@@ -11,6 +11,181 @@
+			DirtRoad: 100
+			Rough: 80
+			Tiberium: 90
+				PathingCost: 90
```
Same here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -428,29 +132,11 @@ public void SetLocation(CPos from, SubCell fromSub, CPos to, SubCell toSub)
-					!string.IsNullOrEmpty(Info.JumpjetCondition) && jumpjetToken == ConditionManager.InvalidConditionToken)
-				jumpjetToken = conditionManager.GrantCondition(self, Info.JumpjetCondition);
+			// Tunnel condition is added/removed when starting the transition between layers.
+			// Jumpjet condition is granted at this point, too.
```
Are these comments still appropriate (or needed)? I thought we kind of wanted to make this more abstract.

reaperrr commented:
Updated.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/JumpjetActorLayer.cs:
```diff
@@ -72,28 +72,31 @@ WPos ICustomMovementLayer.CenterOfCell(CPos cell)
 			var terrainType = map.GetTerrainInfo(cell).Type;
-			if (!mi.JumpjetTransitionTerrainTypes.Contains(terrainType) && mi.JumpjetTransitionTerrainTypes.Any())
+			var jli = li as JumpjetLocomotorInfo;
+			if (!jli.JumpjetTransitionTerrainTypes.Contains(terrainType) && jli.JumpjetTransitionTerrainTypes.Any())
```
Minor point that occurs a few times: `as` followed with immediate dereferencing is better served with a normal cast (i.e. `var jli = (JumpjetLocomotorInfo)li;`).

If you're wrong about the type and use a cast, you get an informative InvalidCastException at runtime.
If you're wrong about the type and use `as`, you get a NullReferenceException at runtime, and sadness when we have try and work out exactly what might have been null when we get the stack trace.

reaperrr commented:
> Minor point that occurs a few times: as followed with immediate dereferencing is better served with a normal cast (i.e. ´var jli = (JumpjetLocomotorInfo)li;´). 

Updated, done for both special-case locomotors.

abcdefg30 commented:
Got
```
OpenRA engine version {DEV_VERSION}
Red Alert mod version {DEV_VERSION}
on map 28ae41d4def5caaaaf79f08870dfb08f56d8d7eb (Desert Shellmap by Scott_NZ).
Date: 2018-03-21 11:18:21Z
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.InvalidCastException`: Das Objekt des Typs "OpenRA.Mods.Common.Traits.LocomotorInfo" kann nicht in Typ "OpenRA.Mods.Common.Traits.SubterraneanLocomotorInfo" umgewandelt werden.
   bei OpenRA.Mods.Common.Traits.Mobile.SetVisualPosition(Actor self, WPos pos) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Mods.Common\Traits\Mobile.cs:Zeile 225.
   bei OpenRA.Mods.Common.Traits.Mobile..ctor(ActorInitializer init, MobileInfo info) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Mods.Common\Traits\Mobile.cs:Zeile 155.
   bei OpenRA.Mods.Common.Traits.MobileInfo.Create(ActorInitializer init) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Mods.Common\Traits\Mobile.cs:Zeile 53.
   bei OpenRA.Actor..ctor(World world, String name, TypeDictionary initDict) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\Actor.cs:Zeile 100.
   bei OpenRA.World.CreateActor(Boolean addToWorld, String name, TypeDictionary initDict) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\World.cs:Zeile 252.
   bei OpenRA.World.CreateActor(String name, TypeDictionary initDict) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\World.cs:Zeile 247.
   bei OpenRA.Mods.Common.Traits.SpawnMapActors.WorldLoaded(World world, WorldRenderer wr) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Mods.Common\Traits\World\SpawnMapActors.cs:Zeile 41.
   bei OpenRA.World.LoadComplete(WorldRenderer wr) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\World.cs:Zeile 226.
   bei OpenRA.Game.StartGame(String mapUID, WorldType type) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\Game.cs:Zeile 173.
   bei OpenRA.Game.LoadShellMap() in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\Game.cs:Zeile 474.
   bei OpenRA.Mods.Common.LoadScreens.BlankLoadScreen.StartGame(Arguments args) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Mods.Common\LoadScreens\BlankLoadScreen.cs:Zeile 100.
   bei OpenRA.Game.InitializeMod(String mod, Arguments args) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\Game.cs:Zeile 461.
   bei OpenRA.Game.Initialize(Arguments args) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\Game.cs:Zeile 366.
   bei OpenRA.Game.InitializeAndRun(String[] args) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\Game.cs:Zeile 251.
   bei OpenRA.Program.Main(String[] args) in c:\Users\Lukas\Documents\OpenRA\bleed\OpenRA\OpenRA.Game\Support\Program.cs:Zeile 37.
```
immediately when starting the RA mod.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -536,29 +222,30 @@ public void SetVisualPosition(Actor self, WPos pos)
 			CenterPosition = pos;
 			self.World.UpdateMaps(self, this);
 
+			var locoInfo = (SubterraneanLocomotorInfo)Info.LocomotorInfo;
```
This needs to stay as `as`.

reaperrr commented:
Fixed.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1774,6 +1774,19 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
+					{
+						// Writing a 'proper' upgrade rule that works well would be extremely hard due to inheritance and World involvement,
+						// notifying modders but leaving the actual changes to them is likely the better approach for everyone.
+						Console.WriteLine("Various properties of Mobile have been split off to Locomotor World trait.");
```
Can you please include a #14964 compatible update rule, following the example of `DefineSoundDefaults` to print out more concrete explanations and yaml locations?

pchote replied:
In the `AfterUpdate` you should first yield return a message explaining how to define locomotor types on the world actor, and then a composite message with all the Mobile references that may need to be looked at.

pchote commented on OpenRA.Mods.Common/Pathfinder/PathGraph.cs:
```diff
@@ -97,15 +97,15 @@ public PathGraph(CellInfoLayerPool layerPool, MobileInfo mobileInfo, Actor actor
 		{
 			pooledLayer = layerPool.Get();
 			groundInfo = pooledLayer.GetLayer();
+			locomotorInfo = mobileInfo.LocomotorInfo;
```
This is the only use of `mobileInfo`, so it would be better if we can pass the `LocomotorInfo` in directly instead.

pchote commented on OpenRA.Mods.Common/AI/States/NavyStates.cs:
```diff
@@ -30,12 +30,12 @@ protected Actor FindClosestEnemy(Squad owner)
 			var domainIndex = first.World.WorldActor.Trait<DomainIndex>();
-			var mobileInfo = first.Info.TraitInfo<MobileInfo>();
-			var passable = (uint)mobileInfo.GetMovementClass(first.World.Map.Rules.TileSet);
+			var locomotorInfo = first.Trait<Mobile>().Info.LocomotorInfo;
```
Why change from `first.Info.TraitInfo<MobileInfo>()` to `first.Trait<Mobile>().Info`?

pchote commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -24,6 +24,7 @@ public class FindResources : Activity
 		readonly HarvesterInfo harvInfo;
 		readonly Mobile mobile;
 		readonly MobileInfo mobileInfo;
```
We should be able to remove some of these cached references after updating `PathGraph` to take the `locomotorInfo`.

pchote commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -161,12 +161,13 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
 					{
 						var mobiles = cargo != null ? cargo.Passengers.Select(a =>
 						{
-							var mobileInfo = a.Info.TraitInfoOrDefault<MobileInfo>();
```
I believe this can simplify down to a much more readable and marginally more efficient:
```csharp
var locomotors = cargo.Passengers
	.Select(a => a.Info.TraitInfo<MobileInfo>())
	.Where(m => m != null)
	.Distinct()
	.Select(m => m.LocomotorInfo)
	.ToList();

foreach (var c in transport.World.Map.FindTilesInCircle(destination, dropRange))
{
	if (!aircraft.CanLand(c))
		continue;

	if (!locomotors.All(m => domainIndex.IsPassable(destination, c, m, (uint)m.GetMovementClass(owner.World.Map.Rules.TileSet))))
		continue;

	destination = c;
	break;
}
```

pchote commented on OpenRA.Mods.Common/Traits/World/DomainIndex.cs:
```diff
@@ -29,23 +29,22 @@ public void WorldLoaded(World world, WorldRenderer wr)
 		}
 
-		public bool IsPassable(CPos p1, CPos p2, MobileInfo mi, uint movementClass)
+		public bool IsPassable(CPos p1, CPos p2, LocomotorInfo loco, uint movementClass)
```
The movement class is determined from the `LocomotorInfo`.  Do we have a good reason to force callers to pass this in separately?

reaperrr replied:
Not as far as I can tell, didn't run into any issues, other than that I need to cache `World` here in DomainIndex now, to be able to pass `world.Map.Rules.TileSet` to `GetMovementClass`.

pchote commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -107,13 +107,13 @@ void INotifyParachute.OnLanded(Actor self, Actor ignore)
 			{
 				// Mobile is (currently) the only trait that supports crushing
-				var mi = a.Info.TraitInfoOrDefault<MobileInfo>();
+				var mi = a.TraitOrDefault<Mobile>();
```
Why change from `TraitInfoOrDefault<MobileInfo>()` to `TraitOrDefault<Mobile>().Info`?  If we want to keep this, then the `mi` should change to `mobile`.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -536,29 +222,30 @@ public void SetVisualPosition(Actor self, WPos pos)
+			if (locoInfo == null)
+				return;
+
 			// HACK: The submerging conditions must be applied part way through a move, and this is the only method that gets called
```
Could we extract this out by createing an `INotifyVisualPositionChanged` that is implemented by the `GrantConditionOnSubterraneanLayer`?

pchote reviewed (commented):
A few more thoughts, but I still haven't made it through all the changes yet!

reaperrr committed:
Locomotor update rule

reaperrr committed:
Don't pass movement class via IsPassable directly

Let IsPassable get that info from the locomotor instead.

reaperrr commented:
Updated, though the update rule is still unfinished, I'll try to get that done tomorrow.

A note about the movement class `uint` needed for `DomainIndex.IsPassable`:
I wonder if it would be possible (and if yes, more efficient) to cache the uint directly in `LocomotorInfo` via `RulesetLoaded`, then we could drop the `world` caching and `TileSet` look-up in `IsPassable` entirely.

I tried locally, but `Ruleset.TileSet` doesn't seem to work.

pchote commented:
Can you please squash the fixup commits, except for "Don't pass movement class via IsPassable directly" into "Split Locomotor trait from Mobile"?

I plan on writing a smarter update rule for you in the near future so that it can report line numbers that modders should look at.

reaperrr commented:
> Can you please squash the fixup commits, except for "Don't pass movement class via IsPassable directly" into "Split Locomotor trait from Mobile"? This also needs a rebase now.

Both done.

pchote commented:
56ae93b1179ce8b559b22525aabe37c8fa807f15 has a more comprehensive update rule, but it will need to be tested and ideally used as a base for the changes to the default mods.

reaperrr commented:
> 56ae93b has a more comprehensive update rule, but it will need to be tested and ideally used as a base for the changes to the default mods.

Updated.
I added & tested (and fixed style issues on) the update rule and applied it on TS, the automated part works and the manual instructions are fine in my eyes.

I refrained from re-doing the shipping mods though, since they'd have consisted of only manual changes anyway, so it would have meant doing the exact same stuff a 2nd time, that's just too much wasted time for my taste.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnLayer.cs:
```diff
@@ -0,0 +1,73 @@
+		public readonly string Condition = null;
+	}
+
+	public abstract class GrantConditionOnLayer : ConditionalTrait<GrantConditionOnLayerInfo>, INotifyCustomLayerChanged
```
Please declare this as:
```csharp
public abstract class GrantConditionOnLayer<InfoType> : ConditionalTrait<InfoType>, INotifyCustomLayerChanged where InfoType : ConditionalTraitInfo
```

and then change `GrantConditionOnLayerInfo` to `InfoType` in the constructor below.

This then allows you to declare the correct *Info type from the subclasses using e.g.:
```csharp
public class GrantConditionOnSubterraneanLayer : GrantConditionOnLayer<GrantConditionOnSubterraneanLayerInfo>, INotifyVisualPositionChanged
```
and you can then drop the `localInfo` declarations.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -145,240 +52,38 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
 		{
-			if (!world.Map.Contains(cell))
-				return int.MaxValue;
+			var locoInfos = rules.Actors["world"].TraitInfos<LocomotorInfo>();
```
please call this `locomotorInfos`.  Shortening it doesn't help readability.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -159,21 +159,19 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
-							return new Pair<MobileInfo, uint>(mobileInfo, (uint)mobileInfo.GetMovementClass(a.World.Map.Rules.TileSet));
-						}) : new Pair<MobileInfo, uint>[0];
+						var locomotors = cargo.Passengers
+							.Select(a => a.Info.TraitInfo<MobileInfo>())
```
This should be a `TraitInfoOrDefault` - we then filter out any nulls on the next line.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnSubterraneanLayer.cs:
```diff
@@ -0,0 +1,85 @@
+		protected override void Created(Actor self)
+		{
+			var mobileInfo = self.Info.TraitInfo<MobileInfo>();
+			var li = (SubterraneanLocomotorInfo)mobileInfo.LocomotorInfo;
```
Can we please add a `RulesetLoaded` check on the info to throw a `YamlException` if `mobileInfo.LocomotorInfo` isn't a `SubterraneanLocomotorInfo`?

reaperrr replied:
Done (on *JumpjetLayer as well).

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -145,240 +52,38 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
 
-			var index = cell.Layer == 0 ? world.Map.GetTerrainIndex(cell) :
-				world.GetCustomMovementLayers()[cell.Layer].GetTerrainIndex(cell);
+			LocomotorInfo = locomotorInfos.Single(li => li.Name == Locomotor);
```
We can save 2/3 of the collection enumerations by changing this to:
```csharp
LocomotorInfo = locomotorInfos.SingleOrDefault(li => li.Name == Locomotor);
if (LocomotorInfo == null)
{
	if (locomotorInfos.Count(li => li.Name == Locomotor) == 0)
		throw new YamlException("A locomotor named '{0}' doesn't exist.".F(Locomotor));
	else
		throw new YamlException("There is more than one locomotor named '{0}'.".F(Locomotor));
}
```

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -684,12 +352,13 @@ public bool IsLeavingCell(CPos location, SubCell subCell = SubCell.Any)
 		public SubCell GetAvailableSubCell(CPos a, SubCell preferredSubCell = SubCell.Any, Actor ignoreActor = null, bool checkTransientActors = true)
 		{
-			return Info.GetAvailableSubCell(self.World, self, a, preferredSubCell, ignoreActor, checkTransientActors ? CellConditions.All : CellConditions.None);
+			return Info.LocomotorInfo
```
Minor style nit: It would look a bit nicer if we split out a `var cellConditions = checkTransientActors ? CellConditions.All : CellConditions.None` instead of breaking this statement over two lines.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -714,15 +383,16 @@ public void EnteringCell(Actor self)
-			// and that condition guarantees that this only runs when the unit has finished landing.
-			if (fromCell.Layer != CustomMovementLayerType.Jumpjet && toCell.Layer != CustomMovementLayerType.Jumpjet && jumpjetToken != ConditionManager.InvalidConditionToken)
-				jumpjetToken = conditionManager.RevokeCondition(self, jumpjetToken);
+			// and that condition guarantees that this only runs when the unit has finished any potential layer transition.
```
This comment doesn't make sense to me anymore.  Was I referring to the if-statement condition or the condition-manager condition?  If the latter then this line of the comment should be removed.

pchote commented on OpenRA.Mods.Common/Traits/World/DomainIndex.cs:
```diff
@@ -23,31 +23,32 @@ class DomainIndexInfo : TraitInfo<DomainIndex> { }
 
 			// HACK: Workaround until we can generalize movement classes
-			if (mi.Subterranean || mi.Jumpjet)
+			if (li is SubterraneanLocomotorInfo || li is JumpjetLocomotorInfo)
```
This PR has generalized movement classes enough to remove this hack:
* Add a `public virtual bool DisableDomainPassabilityCheck { get { return false; } }` property to `LocomotorInfo`
* Override it to return false in `SubterraneanLocomotorInfo` and `JumpjetLocomotorInfo`
* Change this to `if (li.DisableDomainPassabilityCheck)` and remove the `// HACK` comment.

pchote commented on OpenRA.Mods.Common/Traits/World/SubterraneanLocomotor.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Used by Mobile. Attach these to the world actor. You can have multiple variants by adding @suffixes.")]
```
Can we tweak this so that it explains it is part of the subterranean movement logic?

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/DefineLocomotors.cs:
```diff
@@ -0,0 +1,162 @@
+			get
+			{
+				return "A large number of properties have been moved from the actor-level Mobile trait\n" +
+					"to either world-level Locomotor traits or new actor-level GrantCondition* traits.";
```
Can we please add a line explaining what the rule actually does? Something like "Conditions for subterranean and jumpjet behaviours are migrated, and affected Mobile traits are listed for inspection."

pchote reviewed (commented):
This is looking pretty good to me now.  I have one hopefully final round of minor nits then I think this will be ready to go.  Also needs a rebase.

reaperrr committed:
Replace DomainIndex passability hack with Locomotor boolean

reaperrr commented:
Updated and rebased.

pchote reviewed (dismissed):
LGTM :+1:  Any regressions that may be hiding inside here will hopefully be shaken out during playtesting.

abcdefg30 commented on OpenRA.Mods.Common/Lint/CheckLocomotorReferences.cs:
```diff
@@ -0,0 +1,51 @@
+		void CheckLocomotors(ActorInfo actorInfo, Action<string> emitError, Ruleset rules, string locomotor)
+		{
+			var worldActor = rules.Actors["world"];
+			var locomotorInfos = worldActor.TraitInfos<LocomotorInfo>();
```
Minor: You could pull this out of all the foreach loops probably, as this isn't changing during the lint run?

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/PathFinder.cs:
```diff
@@ -62,20 +63,17 @@ public PathFinder(World world)
-			{
-				var passable = mi.GetMovementClass(world.Map.Rules.TileSet);
-				if (!domainIndex.IsPassable(source, target, mi, (uint)passable))
+				if (!domainIndex.IsPassable(source, target, li))
```
Neglectible: I'd be nice if this was combined into the if check above.

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/DefineLocomotors.cs:
```diff
@@ -0,0 +1,164 @@
+					var grantNode = new MiniYamlNode("GrantConditionOnTunnelLayer", "");
+					grantNode.AddNode("Condition", tunnelConditionNode.Value.Value);
+					addNodes.Add(grantNode);
+					mobileNode.Value.Nodes.Remove(tunnelConditionNode);
```
This applies to the entire file: You should use the respective `Remove` and `Add` methods instead of using `.Value.Nodes`.

reaperrr replied:
Done, but only In places I was confident about (this rule was pretty much written by @pchote).

abcdefg30 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -536,31 +225,12 @@ public void SetVisualPosition(Actor self, WPos pos)
-					if (!string.IsNullOrEmpty(Info.SubterraneanTransitionSound))
-						Game.Sound.Play(SoundType.World, Info.SubterraneanTransitionSound);
+			// The first time SetVisualPosition is called is in the constructor before creation, so we need a null check here as well
+			if (notifyVisualPositionChanged == null || !notifyVisualPositionChanged.Any())
```
Neglectible: I don't think the `.Any` check is needed. When the collection is empty, we'll just loop zero times.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -145,240 +52,40 @@ IEnumerable<object> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, ActorP
-			internal readonly TerrainInfo[] TerrainInfos;
-			internal WorldMovementInfo(World world, MobileInfo info)
+			var locomotorInfos = rules.Actors["world"].TraitInfos<LocomotorInfo>();
+			LocomotorInfo = locomotorInfos.SingleOrDefault(li => li.Name == Locomotor);
```
The part below won't work. `SingleOrDefault` will already crash when there are multiple locomotors with the same name.

reaperrr replied:
I rewrote the entire check. SingleOrDefault didn't return a meaningful error, so I changed it to FirstOrDefault and check the Count later instead.

abcdefg30 commented:
Needs a rebase. Looks good to me otherwise. 👍 / ✅

reaperrr commented:
Rebased and updated.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0fd8feccc14f49b137f8c0582fee977a447ba322...eec4c09115dba0dbda5e689970b3c82dac031301)
